src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -0,0 +1,232 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonReaderHelper
+    {
+        // Reject any invalid UTF-8 data rather than silently replacing.
+        public static readonly UTF8Encoding s_utf8Encoding = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false, throwOnInvalidBytes: true);
+
+        // TODO: Similar to escaping, replace the unescaping logic with publicly shipping APIs from https://github.com/dotnet/corefx/issues/33509
+        public static string GetUnescapedString(ReadOnlySpan<byte> utf8Source, int idx)
+        {
+            byte[] unescapedArray = null;
+
+            Span<byte> utf8Unescaped = utf8Source.Length <= JsonConstants.StackallocThreshold ?
+                stackalloc byte[utf8Source.Length] :
+                (unescapedArray = ArrayPool<byte>.Shared.Rent(utf8Source.Length));
+
+            Unescape(utf8Source, utf8Unescaped, idx, out int written);
+            Debug.Assert(written > 0);
+
+            utf8Unescaped = utf8Unescaped.Slice(0, written);
+            Debug.Assert(!utf8Unescaped.IsEmpty);
+
+            string utf8String = TranscodeHelper(utf8Unescaped);
+
+            if (unescapedArray != null)
+            {
+                utf8Unescaped.Clear();
+                ArrayPool<byte>.Shared.Return(unescapedArray);
+            }
+
+            return utf8String;
+        }
+
+        public static string TranscodeHelper(ReadOnlySpan<byte> utf8Unescaped)
+        {
+            try
+            {
+#if BUILDING_INBOX_LIBRARY
+                return s_utf8Encoding.GetString(utf8Unescaped);
+#else
+                if (utf8Unescaped.IsEmpty)
+                {
+                    return string.Empty;
+                }
+                unsafe
+                {
+                    fixed (byte* bytePtr = utf8Unescaped)
+                    {
+                        return s_utf8Encoding.GetString(bytePtr, utf8Unescaped.Length);
+                    }
+                }
+#endif
+            }
+            catch (DecoderFallbackException)
```

Comment:
Ideally we would have the `OperationStatus` based `ToUtf16` API and then rather than catching this exception, we would just check the status and throw the correct exception.

Reply:
Based on offline feedback from @bartonjs, wrapping the `DecoderFallbackException` within `InvalidOperationException`.